File: AgoraChat/Code/Controllers/Groups/ACDContainerSearchTableViewController+GroupMemberList.m
//
//  ACDContainerSearchTableViewController+GroupMemberList.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/31.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDContainerSearchTableViewController+GroupMemberList.h"
#import <objc/runtime.h>

typedef void(^actionBlock)();

#define kActionAdminKey   @"ActionAdminKey"
#define kActionUnAdminKey @"ActionUnAdminKey"

#define kActionMuteKey    @"ActionMuteKey"
#define kActionUnMuteKey  @"ActionUnMuteKey"

#define kActionBlockKey   @"ActionBlockKey"
#define kActionUnBlockKey @"ActionUnBlockKey"

#define kActionRemoveFromGroupKey @"ActionRemoveFromGroupKey"

static NSString *selectedUserIdKey = @"selectedUserId";
static NSString *groupIdKey = @"groupId";

@interface ACDContainerSearchTableViewController ()

@end

@implementation ACDContainerSearchTableViewController (GroupMemberList)

- (void)actionSheetWithUserId:(NSString *)userId
               memberListType:(ACDGroupMemberListType)memberListType
                        group:(AgoraChatGroup *)group {
    //if selected user is currentUsernam, than do nothing
    if ([userId isEqualToString:AgoraChatClient.sharedClient.currentUsername]) {
        return;
    }
    
    //admin can not opertion admin
    if (group.permissionType == AgoraChatGroupPermissionTypeAdmin) {
        BOOL isAdminUserId = [group.adminList containsObject:userId];
        if (isAdminUserId) {
            return;
        }
    }
    
    if (group.permissionType == AgoraChatGroupPermissionTypeMember) {
        return;
    }
    
    self.selectedUserId = userId;
    self.groupId = group.groupId;
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:userId message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    if (group.permissionType == AgoraChatGroupPermissionTypeOwner) {
        BOOL isAdmin = [group.adminList containsObject:userId];
        NSArray *actions = [self ownerWithMemberListType:memberListType selectedIsAdmin:isAdmin alertController:alertController];
        for (UIAlertAction *action in actions) {
            [alertController addAction:action];
        }
        
    }
    
    if (group.permissionType == AgoraChatGroupPermissionTypeAdmin) {
        NSArray *actions = [self adminWithMemberListType:memberListType alertController:alertController];
        
        for (UIAlertAction *action in actions) {
            [alertController addAction:action];
        }
    }
    
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
    }]];
    
    [self presentViewController:alertController animated:YES completion:nil];
    return;

}
   

-(NSArray *)ownerWithMemberListType:(ACDGroupMemberListType)memberListType
                    selectedIsAdmin:(BOOL)selectedIsAdmin
                    alertController:(UIAlertController *)alertController {
    
    NSMutableArray *actionArray = NSMutableArray.new;
    NSDictionary *actionDic = [self alertActionDics];
    if (memberListType == ACDGroupMemberListTypeALL) {
        if (selectedIsAdmin) {
            [actionArray addObject:actionDic[kActionUnAdminKey]];
            [actionArray addObject:actionDic[kActionMuteKey]];
            [actionArray addObject:actionDic[kActionBlockKey]];
            [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
    
        }else {
            [actionArray addObject:actionDic[kActionAdminKey]];
            [actionArray addObject:actionDic[kActionMuteKey]];
            [actionArray addObject:actionDic[kActionBlockKey]];
            [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
        }
    }
    
    if (memberListType == ACDGroupMemberListTypeBlock) {
        [actionArray addObject:actionDic[kActionUnBlockKey]];
        [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
    }

    if (memberListType == ACDGroupMemberListTypeMute) {
        [actionArray addObject:actionDic[kActionUnMuteKey]];
        [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
    }

    return [actionArray copy];
    
}


-(NSArray *)adminWithMemberListType:(ACDGroupMemberListType)memberListType
               alertController:(UIAlertController *)alertController {
    
    NSMutableArray *actionArray = NSMutableArray.new;
    NSDictionary *actionDic = [self alertActionDics];
    if (memberListType == ACDGroupMemberListTypeALL) {
        [actionArray addObject:actionDic[kActionMuteKey]];
        [actionArray addObject:actionDic[kActionBlockKey]];
        [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
    }
    
    if (memberListType == ACDGroupMemberListTypeBlock) {
        [actionArray addObject:actionDic[kActionUnBlockKey]];
        [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
    }

    if (memberListType == ACDGroupMemberListTypeMute) {
        [actionArray addObject:actionDic[kActionUnMuteKey]];
        [actionArray addObject:actionDic[kActionRemoveFromGroupKey]];
    }

    return [actionArray copy];

}

- (NSDictionary *)alertActionDics {
        NSMutableDictionary *alertActionDics = NSMutableDictionary.new;
  
    UIAlertAction *makeAdminAction = [UIAlertAction alertActionWithTitle:@"Make Admin" iconImage:ImageWithName(@"admin") textColor:TextLabelBlackColor alignment:NSTextAlignmentLeft completion:^{
        [self makeAdmin];
    }];


    UIAlertAction *makeMuteAction = [UIAlertAction alertActionWithTitle:@"Mute" iconImage:ImageWithName(@"mute") textColor:TextLabelBlackColor alignment:NSTextAlignmentLeft completion:^{
        [self makeMute];
    }];
    
    UIAlertAction *makeBlockAction = [UIAlertAction alertActionWithTitle:@"Move to Blocked List" iconImage:ImageWithName(@"blocked") textColor:TextLabelBlackColor alignment:NSTextAlignmentLeft completion:^{
        [self makeBlock];
    }];
    
    
    UIAlertAction *makeRemoveGroupAction = [UIAlertAction alertActionWithTitle:@"Remove From Group" iconImage:ImageWithName(@"remove") textColor:TextLabelPinkColor alignment:NSTextAlignmentLeft completion:^{
        [self makeRemoveGroup];
    }];

        
    UIAlertAction *makeUnAdminAction = [UIAlertAction alertActionWithTitle:@"Remove as Admin" iconImage:ImageWithName(@"remove_admin") textColor:TextLabelBlackColor alignment:NSTextAlignmentLeft completion:^{
        [self unAdmin];
    }];


    UIAlertAction *makeUnMuteAction = [UIAlertAction alertActionWithTitle:@"Unmute" iconImage:ImageWithName(@"Unmute") textColor:TextLabelBlackColor alignment:NSTextAlignmentLeft completion:^{
        [self unMute];
    }];
    
    
    UIAlertAction *makeUnBlockAction = [UIAlertAction alertActionWithTitle:@"Remove from Blocked List" iconImage:ImageWithName(@"Unblock") textColor:TextLabelBlackColor alignment:NSTextAlignmentLeft completion:^{
        [self unBlock];
    }];
    
    alertActionDics[kActionAdminKey] = makeAdminAction;
    alertActionDics[kActionMuteKey] =  makeMuteAction;
    alertActionDics[kActionBlockKey] = makeBlockAction;
    alertActionDics[kActionUnAdminKey] = makeUnAdminAction;
    alertActionDics[kActionUnMuteKey] = makeUnMuteAction;
    alertActionDics[kActionUnBlockKey] = makeUnBlockAction;
    alertActionDics[kActionRemoveFromGroupKey] = makeRemoveGroupAction;

    return alertActionDics;
}

//- (UIAlertAction* )alertActionWithTitle:(NSString *)title
//                             completion:(actionBlock)completion {
//    UIAlertAction* alertAction = [UIAlertAction actionWithTitle:title style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
//        if (completion) {
//            completion();
//        }
//    }];
//    return alertAction;
//}


#pragma mark actions
- (void)makeAdmin {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager addAdmin:self.selectedUserId toGroup:self.groupId error:&error];
    [self handleActionWithMemberListType:ACDGroupMemberListTypeAdmin responseError:error];
}




- (void)unAdmin {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager removeAdmin:self.selectedUserId fromGroup:self.groupId error:&error];
    
    [self handleActionWithMemberListType:ACDGroupMemberListTypeAdmin responseError:error];
}


- (void)makeMute {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager muteMembers:@[self.selectedUserId] muteMilliseconds:-1 fromGroup:self.groupId error:&error];
    
    [self handleActionWithMemberListType:ACDGroupMemberListTypeMute responseError:error];

}

- (void)unMute {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager unmuteMembers:@[self.selectedUserId] fromGroup:self.groupId error:&error];
    [self handleActionWithMemberListType:ACDGroupMemberListTypeMute responseError:error];

}


- (void)makeBlock {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager blockOccupants:@[self.selectedUserId] fromGroup:self.groupId error:&error];
    [self handleActionWithMemberListType:ACDGroupMemberListTypeBlock responseError:error];
}

- (void)unBlock {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager unblockOccupants:@[self.selectedUserId] forGroup:self.groupId error:&error];
    [self handleActionWithMemberListType:ACDGroupMemberListTypeBlock responseError:error];
}


- (void)makeRemoveGroup {
    AgoraChatError *error = nil;
    [[AgoraChatClient sharedClient].groupManager removeOccupants:@[self.selectedUserId] fromGroup:self.groupId error:&error];
    [self handleActionWithMemberListType:ACDGroupMemberListTypeALL responseError:error];
}

- (void)handleActionWithMemberListType:(ACDGroupMemberListType)memberListType
                         responseError:(AgoraChatError *)error {
    if (self.isSearchState) {
        [self cancelSearchState];
    }
    
    if (error == nil) {
        [[NSNotificationCenter defaultCenter] postNotificationName:KACD_REFRESH_GROUP_MEMBER object:@{kACDGroupId:self.groupId,kACDGroupMemberListType:@(memberListType)}];
    }else {
        [self showHint:error.errorDescription];
    }
}


#pragma mark getter and setter
- (void)setSelectedUserId:(NSString *)selectedUserId {
    objc_setAssociatedObject(self, &selectedUserIdKey, selectedUserId, OBJC_ASSOCIATION_COPY);
}

- (NSString *)selectedUserId {
    return objc_getAssociatedObject(self, &selectedUserIdKey);
}


- (void)setGroupId:(NSString *)groupId {
    objc_setAssociatedObject(self, &groupIdKey, groupId, OBJC_ASSOCIATION_COPY);
}

- (NSString *)groupId {
    return objc_getAssociatedObject(self, &groupIdKey);
}

@end

#undef kActionAdminKey
#undef kActionUnAdminKey

#undef kActionMuteKey
#undef kActionUnMuteKey

#undef kActionBlockKey
#undef kActionUnBlockKey

#undef kActionRemoveFromGroupKey

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Groups/ACDContainerSearchTableViewController+GroupMemberList.swift
//
//  ACDContainerSearchTableViewController+GroupMemberList.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/21.
//

import UIKit

private var ActionAdminKey = "ActionAdminKey"
private var ActionUnAdminKey = "ActionUnAdminKey"

private var ActionMuteKey = "ActionMuteKey"
private var ActionUnMuteKey = "ActionUnMuteKey"

private var ActionBlockKey = "ActionBlockKey"
private var ActionUnBlockKey = "ActionUnBlockKey"

private var ActionRemoveFromGroupKey = "ActionRemoveFromGroupKey"

private var SelectedUserIdKey = "selectedUserId"
private var GroupIdKey = "groupId"

extension ACDContainerSearchTableViewController where T == [AgoraUserModel] {
    enum GroupMemberShowType: Int {
        case all = 0
        case admin
        case mute
        case block
        case white
    }
    
    private var selectedUserId: String? {
        set {
            let key = withUnsafePointer(to: &SelectedUserIdKey) { UnsafeRawPointer($0) }
            objc_setAssociatedObject(self, key, newValue, .OBJC_ASSOCIATION_COPY)
        }
        get {
            let key = withUnsafePointer(to: &SelectedUserIdKey) { UnsafeRawPointer($0) }
            return objc_getAssociatedObject(self, key) as? String
        }
    }
    
    private var groupId: String? {
        set {
            let key = withUnsafePointer(to: &GroupIdKey) { UnsafeRawPointer($0) }
            objc_setAssociatedObject(self, key, newValue, .OBJC_ASSOCIATION_COPY)
        }
        get {
            let key = withUnsafePointer(to: &GroupIdKey) { UnsafeRawPointer($0) }
            return objc_getAssociatedObject(self, key) as? String
        }
    }
    
    func showActionSheet(userId: String, showType: ACDContainerSearchTableViewController.GroupMemberShowType, group: AgoraChatGroup) {
        if userId == AgoraChatClient.shared().currentUsername {
            return
        }
        //admin can not opertion admin
        if group.permissionType == .admin {
            if group.adminList.contains(userId) {
                return
            }
        }
        if group.permissionType == .member {
            return
        }
        
        self.selectedUserId = userId
        self.groupId = group.groupId
        
        let alertController = UIAlertController(title: userId, message: nil, preferredStyle: .actionSheet)
        if group.permissionType == .owner {
            self.addOwnerActions(type: showType, selectedIsAdmin: group.adminList.contains(userId), isMited: group.muteList.contains(userId), alertController: alertController)
        } else if group.permissionType == .admin {
            self.addAdminActions(type: showType, isMuted: group.muteList.contains(userId), alertController: alertController)
        }
        
        alertController.addAction(UIAlertAction(title: LocalizedString.Cancel, style: .cancel))
        self.present(alertController, animated: true)
    }
    
    private func addOwnerActions(type: GroupMemberShowType, selectedIsAdmin: Bool, isMited: Bool, alertController: UIAlertController) {
        let map = self.alertActionMap
        if type == .all {
            if selectedIsAdmin {
                alertController.addAction(map[ActionUnAdminKey]!)
            } else {
                alertController.addAction(map[ActionAdminKey]!)
            }
            alertController.addAction(map[ActionBlockKey]!)
            if isMited {
                alertController.addAction(map[ActionUnMuteKey]!)
            } else {
                alertController.addAction(map[ActionMuteKey]!)
            }
            alertController.addAction(map[ActionRemoveFromGroupKey]!)
        } else if type == .block {
            alertController.addAction(map[ActionUnBlockKey]!)
            alertController.addAction(map[ActionRemoveFromGroupKey]!)
        } else if type == .mute {
            alertController.addAction(map[ActionUnMuteKey]!)
            alertController.addAction(map[ActionRemoveFromGroupKey]!)
        }
    }
    
    private func addAdminActions(type: GroupMemberShowType, isMuted: Bool, alertController: UIAlertController) {
        let map = self.alertActionMap
        if type == .all {
            if isMuted {
                alertController.addAction(map[ActionUnMuteKey]!)
            } else {
                alertController.addAction(map[ActionMuteKey]!)
            }
            alertController.addAction(map[ActionBlockKey]!)
            alertController.addAction(map[ActionRemoveFromGroupKey]!)
        } else if type == .block {
            alertController.addAction(map[ActionUnBlockKey]!)
            alertController.addAction(map[ActionRemoveFromGroupKey]!)
        } else if type == .mute {
            alertController.addAction(map[ActionUnMuteKey]!)
            alertController.addAction(map[ActionRemoveFromGroupKey]!)
        }
    }
    
    private var alertActionMap: [String: UIAlertAction] {
        get {
            var map: [String: UIAlertAction] = [:]
            map[ActionAdminKey] = UIAlertAction(title: "Make Admin".localized, iconImage: UIImage(named: "admin"), textColor: .black, alignment: .left, completion: { _ in
                self.makeAdmin()
            })
            map[ActionMuteKey] = UIAlertAction(title: "Mute".localized, iconImage: UIImage(named: "mute"), textColor: .black, alignment: .left, completion: { _ in
                self.makeMute()
            })
            map[ActionBlockKey] = UIAlertAction(title: "Block".localized, iconImage: UIImage(named: "blocked"), textColor: .black, alignment: .left, completion: { _ in
                self.makeBlock()
            })
            map[ActionUnAdminKey] = UIAlertAction(title: "Remove as Admin".localized, iconImage: UIImage(named: "remove_admin"), textColor: .black, alignment: .left, completion: { _ in
                self.unAdmin()
            })
            map[ActionUnMuteKey] = UIAlertAction(title: "Unmute".localized, iconImage: UIImage(named: "Unmute"), textColor: .black, alignment: .left, completion: { _ in
                self.unMute()
            })
            map[ActionUnBlockKey] = UIAlertAction(title: "Unblock".localized, iconImage: UIImage(named: "Unblock"), textColor: .black, alignment: .left, completion: { _ in
                self.unBlock()
            })
            map[ActionRemoveFromGroupKey] = UIAlertAction(title: "Remove From Group".localized, iconImage: UIImage(named: "remove"), textColor: .black, alignment: .left, completion: { _ in
                self.makeRemoveGroup()
            })
            return map
        }
    }
    
    private func makeAdmin() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.addAdmin(selectedUserId, toGroup: groupId, completion: { _, error in
            self.handleAction(type: .admin, error: error)
        })
    }

    private func unAdmin() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.removeAdmin(selectedUserId, fromGroup: groupId, completion: { _, error in
            self.handleAction(type: .admin, error: error)
        })
    }
    
    private func makeMute() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.muteMembers([selectedUserId], muteMilliseconds: -1, fromGroup: groupId, completion: { _, error in
            self.handleAction(type: .mute, error: error)
        })
    }

    private func unMute() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.unmuteMembers([selectedUserId], fromGroup: groupId, completion: { _, error in
            self.handleAction(type: .mute, error: error)
        })
    }

    private func makeBlock() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.blockMembers([selectedUserId], fromGroup: groupId, completion: { _, error in
            self.handleAction(type: .block, error: error)
        })
    }

    private func unBlock() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.unblockMembers([selectedUserId], fromGroup: groupId, completion: { _, error in
            self.handleAction(type: .block, error: error)
        })
    }
    
    private func makeRemoveGroup() {
        guard let selectedUserId = self.selectedUserId, let groupId = self.groupId else {
            return
        }
        AgoraChatClient.shared().groupManager?.removeMembers([selectedUserId], fromGroup: groupId, completion: { _, error in
            self.handleAction(type: .all, error: error)
        })
    }

    private func handleAction(type: GroupMemberShowType, error: AgoraChatError?) {
        if self.isSearchState {
            self.cancelSearchState()
        }
        if let error = error {
            self.showHint(error.errorDescription)
        } else {
            NotificationCenter.default.post(name: GroupMemberChangedNotification, object: [
                "kACDGroupId": self.groupId ?? "",
                "kACDGroupMemberListType": type
            ])
        }
    }
}
